File: 2022/typescript/02.ts
import {readFileSync} from 'fs'

enum Move {
  Rock,
  Paper,
  Scissors
}

enum Outcome {
  Lose,
  Draw,
  Win
}

const input = readFileSync('./2022/02.input', 'utf8')
  .split(/\n/)
  .slice(0, -1)
  .map(x => {
    const [a, b] = x.split(/\s/)
    return [a.charCodeAt(0) - 'A'.charCodeAt(0), b.charCodeAt(0) - 'X'.charCodeAt(0)]
  })

function play(a: Move, b: Move): Outcome {
  if (a === b) {
    return Outcome.Draw
  } else if ((a + 1) % 3 === b) {
    return Outcome.Win
  } else {
    return Outcome.Lose
  }
}

const first = input.map(([a, b]) => play(a, b) * 3 + b + 1).reduce((a, e) => a + e)
console.log(first)

function move(a: Move, outcome: Outcome): Move {
  switch (outcome) {
    case Outcome.Win:
      return (a + 1) % 3
    case Outcome.Draw:
      return a
    case Outcome.Lose:
      return (a + 2) % 3
  }
}

const second = input.map(([a, outcome]) => move(a, outcome) + 1 + outcome * 3).reduce((a, e) => a + e)
console.log(second)

File: go/2022/02/02.go
package main

import (
	"fmt"
	"os"
	"strings"
)

const (
	Lose int = iota
	Draw
	Win
)

func play(a int, b int) int {
	if a == b {
		return Draw
	} else if ((a + 1) % 3) == b {
		return Win
	} else {
		return Lose
	}
}

func move(a int, outcome int) int {
	if outcome == Draw {
		return int(a)
	} else if outcome == Win {
		return (a + 1) % 3
	} else {
		return (a + 2) % 3
	}
}

func main() {
	input, _ := os.ReadFile("../input/2022/02.input")
	lines := strings.Split(strings.Trim(string(input), "\n"), "\n")

	games := make([][2]int, len(lines))
	for i, line := range lines {
		columns := strings.Split(line, " ")
		games[i] = [2]int{int(columns[0][0] - 65), int(columns[1][0] - 88)}
	}

	score := 0
	for _, game := range games {
		score += (game[1] + 1) + (play(game[0], game[1]) * 3)
	}

	fmt.Println(score) // 12586

	score = 0
	for _, game := range games {
		score += (move(game[0], game[1]) + 1) + (game[1] * 3)
	}

	fmt.Println(score) // 13193
}
